### ISSUE TITLE
mpy-cross compile failed on Mac M1 Max

### ISSUE THREAD

**Issue by edwios** [micropython/micropython#8467]
## Cloned a fresh copy from the repo, ran `make -C mpy-cross` and compile failed with **array index refers past the last possible element for an array** error.

**Machine**
MacBook Pro (14-inch, 2021)
Chip Apple M1 Max
Memory 64GB
OS Version 12.2.1

```
uname -a
Darwin MacBook-Pro-2.local 21.3.0 Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T6000 arm64
```

## Steps to reproduce:

```
git clone git@github.com:micropython/micropython.git
cd micropython
make -C mpy-cross 
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
mkdir -p build/genhdr
GEN build/genhdr/mpversion.h
GEN build/genhdr/moduledefs.h
.
.
.
CC ../py/modthread.c
CC ../py/vm.c
../py/vm.c:748:25: error: array index -3 refers past the last possible element for an array in 64-bit address space containing 64-bit (8-byte) elements (max possible 2305843009213693952 elements) [-Werror,-Warray-bounds]
                        sp[-MP_OBJ_ITER_BUF_NSLOTS + 1] = MP_OBJ_NULL;
                        ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
../py/vm.c:272:13: note: array 'sp' declared here
            mp_obj_t *sp = code_state->sp;
            ^
../py/vm.c:749:25: error: array index -2 refers past the last possible element for an array in 64-bit address space containing 64-bit (8-byte) elements (max possible 2305843009213693952 elements) [-Werror,-Warray-bounds]
                        sp[-MP_OBJ_ITER_BUF_NSLOTS + 2] = obj;
                        ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
../py/vm.c:272:13: note: array 'sp' declared here
            mp_obj_t *sp = code_state->sp;
            ^
../py/vm.c:760:25: error: array index -3 refers past the last possible element for an array in 64-bit address space containing 64-bit (8-byte) elements (max possible 2305843009213693952 elements) [-Werror,-Warray-bounds]
                    if (sp[-MP_OBJ_ITER_BUF_NSLOTS + 1] == MP_OBJ_NULL) {
                        ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
../py/vm.c:272:13: note: array 'sp' declared here
            mp_obj_t *sp = code_state->sp;
            ^
../py/vm.c:761:31: error: array index -2 refers past the last possible element for an array in 64-bit address space containing 64-bit (8-byte) elements (max possible 2305843009213693952 elements) [-Werror,-Warray-bounds]
                        obj = sp[-MP_OBJ_ITER_BUF_NSLOTS + 2];
                              ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
../py/vm.c:272:13: note: array 'sp' declared here
            mp_obj_t *sp = code_state->sp;
            ^
4 errors generated.
make: *** [build/py/vm.o] Error 1
```

**Comment by dlech** [comment]
Maybe these should be written using pointer arithmetic instead of array notation? For example:

    sp - MP_OBJ_ITER_BUF_NSLOTS + 1 = MP_OBJ_NULL;

**Comment by dpgeorge** [comment]
Thanks for the report.  I can reproduce this build error on Linux using clang (`make CC=clang` in `mpy-cross` directory).  I have clang 13.0.1.

**Comment by dpgeorge** [comment]
> Maybe these should be written using pointer arithmetic instead of array notation?

This seems to work:
```
*(sp - MP_OBJ_ITER_BUF_NSLOTS + 1) = MP_OBJ_NULL;
```
But this doesn't (!!):
```
*(sp + -MP_OBJ_ITER_BUF_NSLOTS + 1) = MP_OBJ_NULL;
```

**Comment by dpgeorge** [comment]
Fixed by bb70874111dbb246624a68c013e8f1c3245ca0d8
